design/machine-remediation-request.md

```diff
@@ -0,0 +1,142 @@
+To make possible to specify the remediation strategy for different failure cases and for nodes that run on different providers, this proposal introduces **MachineRemediationRequest** controller that should monitor all **MachineRemediationRequest** CR's and apply the remediation strategy according to details specified under the CR.
+When the **MachineRemediationRequest** will apply the remediation strategy, one who is initiating it will be the **MachineHealthCheck** controller. It will monitor nodes, and for each unhealthy node will create **MachineRemediationRequest** CR, if after the specified timeout the **MachineRemediationRequest** will not have the `remediationState` equal to `Succeeded`, the **MachineHealthCheck** controller will notify about the remediation failure, and can try to use different type of the remediation strategy.
+
+#### MachineRemediationRequest spec API
+
+This proposal introduces a new API type: **MachineRemediationRequest**, that makes possible to create simple **MachineRemediationRequest** object with the type of the remediation strategy for the specific machine.
+
+##### Example
+
+```yaml
+apiVersion: machineremediationrequest.metal3.io/v1alpha1
+kind: MachineRemediationRequest
+metadata:
+  name: mrr-test
+  namespace: openshift-machine-api
+spec:
+  type: reboot
+  machineName: test-machine
+```
+
+#### MachineRemediationRequest status API
+
+**MachineRemediationRequest** status will show what the state the remediation operation has and the time when the remediation operation started.
+
+```yaml
+status:
+  remediationState: InProgress
+  remediationStartTime: Thu, 20 Jun 2019 03:38:39 -0400
+```
+
+### Risks and Mitigations
+
+It can introduce some integration complexity between **MachineHealthCheck** and **MachineRemediationRequest** controllers,
+that will make harder to debug and to test things.
+
+## Design Details
+
+### Work Items
+
+* Create new **MachineRemediationRequest** CRD.
+* Write a new controller that will watch **MachineRemediationRequest** CR's.
+* Write functional tests under <https://github.com/openshift/cluster-api-actuator-pkg> repository.
+* Create an operator to deploy the **MachineRemediationRequest** controller.
+* Create integration with the OLM.
+
+### Dependencies
+
+N/A
+
+### Test Plan
+
+As basic tests, it will be enough to create **MachineRemediationRequest** CR's with the reboot type and verify that the machine really rebooted. In the future, we will need to test integration between **MachineHealthCheck** and **MachineRemediationRequest** controllers to verify that our final fencing solution works as expected.
+
+### Upgrade / Downgrade Strategy
+
+Should be managed by OLM.
+
+### Version Skew Strategy
+
+The **MachineRemediationRequest** controller version should be the same as **MachineHealthCheck** version.
```

Comment:
I think this is going to be hard to achieve unless either the implementation is part of the official machine-api-operator codebase (unlikely) or we fork the machine-api-operator codebase (which there is a very strong push to avoid)

Reply:
I did not say it will be easy:) But if we will implement bare-metal-remediator machine-api operator can deploy, similar to other images.